### PR TITLE
backend: support shard row id for local/importer backend

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Support `shard rowid` for table kv encoder

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Side effects

Related changes

Release Note
- Add support for `shard rowid` for local/import backend.

### CODE REVIEW

**Comment by glorv** [comment]
/run-all-tests tidb=release-5.0-rc tikv=release-5.0-rc pd=release-5.0-rc importer=release-5.0-rc

**Comment by glorv** [comment]
/run-all-tests tidb=release-5.0-rc tikv=release-5.0-rc pd=release-5.0-rc importer=release-5.0-rc

**Comment by glorv** [comment]
/run-all-tests tidb=release-5.0-rc tikv=release-5.0-rc pd=release-5.0-rc importer=release-5.0-rc

**Comment by glorv** [comment]
/run-all-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0 importer=release-3.0

**Comment by glorv** [comment]
/run-all-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0 importer=release-3.0

**Comment by glorv** [comment]
/run-all-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0 importer=release-3.0

**Comment by glorv** [comment]
/run-all-tests tidb=release-4.0 tikv=release-4.0 pd=release-4.0 importer=release-4.0

**Comment by glorv** [comment]
/run-all-tests tidb=release-5.0-rc tikv=release-5.0-rc pd=release-5.0-rc importer=release-5.0-rc

**Comment by glorv** [comment]
/run-all-tests tidb=release-5.0-rc tikv=release-5.0-rc pd=release-5.0-rc importer=pr/89

File: lightning/backend/sql2kv.go [hunk 1]
```diff
@@ -332,26 +334,32 @@ func (kvcodec *tableKVEncoder) Encode(
...
-	if common.TableHasAutoRowID(kvcodec.tbl.Meta()) {
+	if common.TableHasAutoRowID(meta) {
+		rowValue := rowID
 		j := columnPermutation[len(cols)]
 		if j >= 0 && j < len(row) {
 			value, err = table.CastValue(kvcodec.se, row[j], extraHandleColumnInfo, false, false)
+			rowValue = value.GetInt64()
 		} else {
+			if meta.ShardRowIDBits > 0 {
+				shard := (int64(fastrand.Uint32()) & (1<<meta.ShardRowIDBits - 1)) << (autoid.RowIDBitLength - meta.ShardRowIDBits - 1)
```
Comment (kennytm) [inline lightning/backend/sql2kv.go:353]: using `fastrand.Uint32()` means the row ID is not reproducible with checkpoint. please use a deterministic seed like `autoRandomBits`.

File: tests/shard_rowid/run.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    run_sql "SELECT count(*) from shard_rowid.shr"
+    check_contains "count(*): 16"
+
+    run_sql "SELECT count(distinct _tidb_rowid & b'000001111111111111111111111111111111111111111111111111111111111') as count FROM shard_rowid.shr"
+    check_contains "count: 16"
+
+    # since we use random to generate the shard bits, with 16 record, there maybe less than 8 distinct value,
+    # but it should be bigger than 4
+    run_sql "SELECT count(distinct _tidb_rowid >> 60) as count from shard_rowid.shr"
+    check_contains_regexp "count: [5-8]"
```
Comment (kennytm) [inline tests/shard_rowid/run.sh:44]: you could avoid introducing `check_contains_regexp` by moving the check into the SQL:

```sh
run_sql 'SELECT count, count between 5 and 8 as correct from (SELECT count(distinct _tidb_rowid >> 60) as count from shard_rowid.shr) _'
check_contains 'correct: 1'
```

**Review by kennytm — COMMENTED** [review]
rest LGTM

File: lightning/backend/sql2kv.go [hunk 3]
```diff
@@ -65,15 +67,27 @@ func NewTableKVEncoder(tbl table.Table, options *SessionOptions) (Encoder, error
...
 		}
+	} else if meta.ShardRowIDBits > 0 {
+		rd := rand.New(rand.NewSource(options.AutoRandomSeed))
+		mask := int64(1)<<meta.ShardRowIDBits - 1
+		shift := autoid.RowIDBitLength - meta.ShardRowIDBits - 1
+		autoIDFn = func(id int64) int64 {
+			rd.Seed(id)
+			shardBits := (int64(rd.Uint32()) & mask) << shift
+			return shardBits | id
+		}
```
Comment (overvenus) [inline lightning/backend/sql2kv.go:90]: How about using TiDBs `OverflowShardBits`?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: lightning/backend/sql2kv.go (hunk 3 above)
Comment (glorv) [reply lightning/backend/sql2kv.go:90]: Seems `OverflowShardBits` used to check the shard row id is overflow. But here we just want to generate a random shardBits that bind with the lightning generated rowid so it's compatible with checkpoint.

**Comment by glorv** [comment]
closed in favor of https://github.com/pingcap/br/pull/757

**Pull request closed without merging**